### PR TITLE
expr/builtin_cast: implement eval for expression

### PR DESCRIPTION
**Pull request by AndreMouche**
@hicqu @BusyJay @XuHuaiyu PTAL

### CODE REVIEW

File: src/coprocessor/dag/expr/mod.rs [hunk 1]
```diff
@@ -423,7 +427,90 @@ impl Expression {
...
+                        });
+                        return Ok(v);
+                    }
+                    Err(Error::UnKnownSignature(_)) => {}
+                    Err(e) => return Err(e),
+                }
+
+                match self.eval_real(ctx, row) {
+                    Ok(d) => {
+                        let v = if let Some(v) = d {
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/mod.rs]: `d.into()` should be ok.

File: src/coprocessor/dag/expr/mod.rs [hunk 2]
```diff
@@ -63,6 +63,10 @@ quick_error! {
             description("column offset not found")
             display("illegal column offset: {}", offset)
         }
+        UnKnownSignature(sig: ScalarFuncSig) {
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/mod.rs]: `K` should be `k`.

File: src/coprocessor/dag/expr/mod.rs [hunk 3]
```diff
@@ -521,34 +586,80 @@ mod test {
...
+            }
+        }
+    }
 
-        let ctx = StatementContext::default();
-        for tt in tests {
-            let expr = Expression::build(tt.0, tt.1, &ctx);
-            assert_eq!(expr.is_ok(), tt.2);
+    #[test]
+    fn test_expression_eval_as_int() {
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/mod.rs]: Should be `test_expression_eval`.

File: src/coprocessor/dag/expr/mod.rs [hunk 4]
```diff
@@ -423,7 +427,68 @@ impl Expression {
...
+                                Datum::I64(v)
+                            }
+                        });
+                        return Ok(v);
+                    }
+                    Err(Error::UnKnownSignature(_)) => {}
+                    Err(e) => return Err(e),
+                }
+
+                match self.eval_real(ctx, row) {
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/mod.rs]: Use `or_else` seems more readable.

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 5]
```diff
@@ -397,7 +397,7 @@ mod test {
             let lhs = datum_expr(tt.1);
             let rhs = datum_expr(tt.2);
 
-            let op = Expression::build(fncall_expr(tt.0, &[lhs, rhs]), 0, &ctx).unwrap();
+            let op = Expression::build(fncall_expr(tt.0, &[lhs, rhs]), &ctx).unwrap();
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/arithmetic.rs]: tests in this file should be updated too.

Comment (BusyJay) [reply src/coprocessor/dag/expr/arithmetic.rs]: And src/coprocessor/dag/expr/math.rs

**Comment by BusyJay** [comment]
LGTM

File: src/coprocessor/dag/expr/mod.rs [hunk 6]
```diff
@@ -443,8 +447,42 @@ impl Expression {
...
+                        if mysql::has_unsigned_flag(self.get_tp().get_flag() as u64) {
+                            Datum::U64(v as u64)
+                        } else {
+                            Datum::I64(v)
+                        }
+                    })
+                });
+
+                let res = filter(eval_int)
+                    .or_else(|| filter(self.eval_real(ctx, row)))
```
Comment (hicqu) [inline src/coprocessor/dag/expr/mod.rs]: we match many times. Would it cause performance lose? @BusyJay

Comment (BusyJay) [reply src/coprocessor/dag/expr/mod.rs]: I think this is only for test purpose.

Comment (BusyJay) [reply src/coprocessor/dag/expr/mod.rs]: Let's merge it first to get things working first. I will figure out a way to reduce the lose.

**Comment by hicqu** [comment]
LGTM.

**Pull request merged**